### PR TITLE
Add org claim validation to the token validation flow.

### PR DESCRIPTION
**Pull request by ShalkiWenushika**
Fixes https://github.com/wso2-enterprise/choreo/issues/15808

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java [hunk 1]
```diff
@@ -74,6 +75,17 @@ public static boolean handleScopeValidation(HashMap<String,Object> message, Sign
                     .map(s -> s.replace(APIConstants.URN_CHOREO + orgId + ":", ""))
                     .toArray(size -> new String[size]);
             oauthTokenInfo.setScopes(scopes);
+            // check whether organization claim value and orgId matches
+            JSONObject orgClaim = signedJWTInfo.getJwtClaimsSet().getJSONObjectClaim("organization");
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java:79]: Need a null check for orgClaim

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java [hunk 2]
```diff
@@ -74,6 +75,17 @@ public static boolean handleScopeValidation(HashMap<String,Object> message, Sign
...
+            String orgUuid = orgClaim.getAsString("uuid");
+            if (orgId == null) {
+                log.error("Organization is not present in the request");
+                return false;
+            }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java:84]: Shall we move this to around line 72 just after `String orgId = (String) message.get(RestApiConstants.ORG_ID);`

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java [hunk 3]
```diff
@@ -151,7 +151,17 @@ private boolean handleScopeValidation(Message message, SignedJWTInfo signedJWTIn
...
                 message.put(ORGANIZATION_HANDLE, orgHandle);
             }
-
+            // check whether organization claim value and orgId matches
+            JSONObject orgClaim = signedJWTInfo.getJwtClaimsSet().getJSONObjectClaim(JwtTokenConstants.ORGANIZATION);
+            String orgUuid = orgClaim.getAsString(JwtTokenConstants.UUID);
+            if (orgId == null) {
+                log.error("Organization is not present in the request");
+                return false;
+            }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java:160]: move up around 148

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java (hunk 1 above)
Comment (ShalkiWenushika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java:83]: Fixed via https://github.com/wso2/carbon-apimgt/pull/11670/commits/a4afd58620f09a8ccb5da7c5822e414948bf6e80

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java (hunk 2 above)
Comment (ShalkiWenushika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/11670/commits/a4afd58620f09a8ccb5da7c5822e414948bf6e80

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java [hunk 4]
```diff
@@ -151,7 +151,17 @@ private boolean handleScopeValidation(Message message, SignedJWTInfo signedJWTIn
             if (orgHandle != null) {
                 message.put(ORGANIZATION_HANDLE, orgHandle);
             }
-
+            // check whether organization claim value and orgId matches
+            JSONObject orgClaim = signedJWTInfo.getJwtClaimsSet().getJSONObjectClaim(JwtTokenConstants.ORGANIZATION);
+            String orgUuid = orgClaim.getAsString(JwtTokenConstants.UUID);
```
[the remark this answers is not in the document]
Comment (ShalkiWenushika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/11670/commits/a4afd58620f09a8ccb5da7c5822e414948bf6e80

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java (hunk 3 above)
Comment (ShalkiWenushika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/11670/commits/a4afd58620f09a8ccb5da7c5822e414948bf6e80

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java [hunk 5]
```diff
@@ -145,13 +145,25 @@ private boolean handleScopeValidation(Message message, SignedJWTInfo signedJWTIn
...
-
+            // check whether organization claim value and orgId matches
+            JSONObject orgClaim = signedJWTInfo.getJwtClaimsSet().getJSONObjectClaim(JwtTokenConstants.ORGANIZATION);
+            if (orgClaim != null) {
+                String orgUuid = orgClaim.getAsString(JwtTokenConstants.UUID);
+                if (orgUuid != null && (!orgId.equals(orgUuid))) {
+                    log.error(String.format("Requested OrgId (%s) and the token's organization uuid (%s) mismatch!", orgId, orgUuid));
+                    return false;
+                }
+            }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java:166]: org id must be there in the JWT. We should not continue if its not there.
9.2.x implementation for a reference:
https://github.com/wso2/carbon-apimgt/blob/9.2.x/components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/utils/JWTUtil.java#L86

Comment (ShalkiWenushika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java:170]: Fixed via https://github.com/wso2/carbon-apimgt/pull/11670/commits/1f32adaf0ff3f026ad4d2a1312b5b1f76a7bfb93

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**